'use warnings' with exceptions

DrHyde opened the pull request:
This makes `use warnings qw(-foo)` equivalent to `use warnings; no warnings qw(foo)`.

Specifically, it splits the list of arguments to `warnings::import` into two classes: "negative imports", ie those beginning with `-`, and "positive imports", that is everything else. The list of positive imports is then processed exactly as before, including inserting an implicit `all` if the list is empty. Then the list of negative imports is processed, by removing the leading `-` from each one and `unimport`ing.

DrHyde committed:
Add -negative import args for 'use warnings'

tonycoz commented:
Squashing the .pm/.pl changes into one commit would improve it, since one is generated from the other.

A couple of more examples, like from your email would improve it too, I think.

DrHyde commented:
> Squashing the .pm/.pl changes into one commit would improve it, since one is generated from the other.

Now that I've pushed changes to github is it possible to rewrite history like that?

As an aside, the generated lib/warnings.pm probably shouldn't be in the repo at all, but should be generated by at `make`-time.

> A couple of more examples, like from your email would improve it too, I think.

OK, I'll add more doco.

demerphq commented:
On Tue, 28 Jan 2020, 17:49 David Cantrell, <notifications@github.com> wrote:

> Squashing the .pm/.pl changes into one commit would improve it, since one
> is generated from the other.
>
> Now that I've pushed changes to github is it possible to rewrite history
> like that?
>
If they were pushed to bleed no. But normally we would push a change like
this to a branch in which case yes.


As an aside, the generated lib/warnings.pm probably shouldn't be in the
> repo at all, but should be generated by at make-time.
>
> A couple of more examples, like from your email would improve it too, I
> think.
>
> OK, I'll add more doco.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Perl/perl5/pull/17500?email_source=notifications&email_token=AAAZ5RZONETN65EAWJ2XQLTQ775TLA5CNFSM4KMFBIB2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKCVMEQ#issuecomment-579163666>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAZ5R2QCYFKMQ6GVHYEZMLQ775TLANCNFSM4KMFBIBQ>
> .
>

tonycoz commented:
> > Squashing the .pm/.pl changes into one commit would improve it, since one is generated from the other.
> 
> Now that I've pushed changes to github is it possible to rewrite history like that?

Yes, just force push to the original branch you created the PR from ( DrHyde:dcantrell/use-warnings-with-exceptions ) and it will replace the commits for the PR.

DrHyde commented:
Right, that's better doco there, and I can't think of any other improvements.

iabyn commented:
On Tue, Jan 28, 2020 at 02:53:13AM -0800, Yves Orton wrote:
> As an aside, the generated lib/warnings.pm probably shouldn't be in the
> repo at all, but should be generated by at make-time.

Tricky, as regen/warnings.pl is a perl script, and also regenerates
warnings.h. So you'd have to build miniperl, run,regen/warnings.pl, then
rebuild miniperl with the new warnings.h.


-- 
You're only as old as you look.

Grinnz commented:
Also please keep in mind that the documentation of such generated modules is only visible on metacpan if it is present in source.

DrHyde commented:
> Also please keep in mind that the documentation of such generated modules is only visible on metacpan if it is present in source.

A very good point!

haarg commented:
Because all negative imports are processed before positive imports, this would prevent something like
```
use warnings qw(all -experimental experimental::smartmatch);
```
from disabling all experimental warnings except smartmatch.

It also doesn't seem to be addressing FATAL warnings.
```
use warnings FATAL => '-newline';
```
would enable all fatal warnings aside from newline, which doesn't seem like expected behavior to me.

DrHyde commented:
> It also doesn't seem to be addressing FATAL warnings.

Deliberately so. I'm primarily concerned with turning warnings *off*. However I have now fixed a bug that I introduced in FATAL handling.

> ```
> use warnings FATAL => '-newline';
> ```
> 
> would enable all fatal warnings aside from newline, which doesn't seem like expected behavior to me.

`use warnings qw(FATAL)`'s behaviour is unchanged - it is still equivalent to `use warnings qw(FATAL all)`. Add a `-something` to that and I think the expected behaviour would be to turn on all warnings, making them fatal (as before), except not turn on warnings for `something`.

If you're thinking that perhaps it should mean the same as `use warnings qw(FATAL all NONFATAL newline)` then ... I disagree. I want the negative warnings syntax to mean *turn off this warning*, as there's already a perfectly good way of twiddling fatality in different directions for different warnings at the same time.

haarg commented:
> `use warnings qw(FATAL)`'s behaviour is unchanged - it is still equivalent to `use warnings qw(FATAL all)`. Add a `-something` to that and I think the expected behaviour would be to turn on all warnings, making them fatal (as before), except not turn on warnings for `something`.

I would not expect it to turn on all fatal warnings.

> 
> If you're thinking that perhaps it should mean the same as `use warnings qw(FATAL all NONFATAL newline)` then ... I disagree. I want the negative warnings syntax to mean _turn off this warning_, as there's already a perfectly good way of twiddling fatality in different directions for different warnings at the same time.

I'd probably expect it to behave the same as `no warnings FATAL => "newline";`, which behaves the same as `no warnings "newline";`.  The interface for fatal warnings is certainly not the best.

DrHyde commented:
> I'd probably expect it to behave the same as `no warnings FATAL => "newline";`, which behaves the same as `no warnings "newline";`. The interface for fatal warnings is certainly not the best.

I intended this new syntax of `use warnings [blahblah] -otherstuff` to mean "turn on blahblah, and turn off otherstuff". If you just want to turn a warning off without also turning anything else on at the same time, then you've got `no warnings` as before. Is there any change you'd suggest in the doco to at least make the expectation match that?

khwilliamson commented:
@DrHyde this needs to be rebased

DrHyde commented:
> @DrHyde this needs to be rebased

Done

Grinnz commented:
For reference, here is the mailing list thread where this was discussed: https://www.nntp.perl.org/group/perl.perl5.porters/2020/01/msg256949.html

Though this ticket has much technical discussion, personally I don't think the complications of interaction with default and new warnings have been fully addressed from a semantics/design standpoint. But I am just an observer.

richardleach commented:
Not an objection from me, just to note that for core, the following will mean "_please do use warnings, but exclude the uninitialized ones_":
`use warnings "-uninitialized";`
Whereas the intention is the opposite in e.g. Mojo land, such as "_please do use Mojo::Base and also include the async/await functionality_":
`use Mojo::Base -async_await;`

Core shouldn't have to slavishly follow CPAN precedence, of course, but the above could confuse some users.

(I don't know if other significant CPAN modules use a similar syntax and, if so, which meaning they confer.)

wchristian commented:
In a similar vein as Richard's comment, i'd also think that a less dangerous syntax would be:

`use warnings -exclude => [qw( foo )];`

ilmari commented:
I would expect the `all` default to only kick in when there are no arguments at all, not when there are no "positive" arguments.
So the equivalent of `use warnings; no warnings 'foo';` would be `use warnings qw(all -foo);`

khwilliamson commented:
I will merge this in 5 more days if there are no further comments.  That totals more than two weeks since the latest change.  @DrHyde , remind me if I (likely) forget

Leont commented:
I think this needs a rebase to filter out the merge commits

DrHyde commented:
Let's hope that rebase didn't irretrievably screw things up!

Grinnz commented on regen/warnings.pl:
```diff
@@ -875,7 +883,10 @@ =head1 SYNOPSIS
-    no warnings "all";
+    no warnings "uninitialized";
+
+    # or equivalent to those last two ...
```
This example is no longer accurate

Grinnz commented on lib/warnings.pm:
```diff
@@ -337,14 +337,22 @@ sub import
+    my @fatal = ();
+    foreach my $warning (@_) {
+        if($warning =~ /^(NON)?FATAL$/) {
+            @fatal = ($warning);
```
This seems to be introducing new behavior where `(NON)?FATAL` will get recognized in places other than the beginning of the import list.

Grinnz replied:
Looking closer at this, I see that it is because `_bits` is now called individually for each argument. I am not perfectly confident in the robustness of this logic but it seems correct.

DrHyde replied:
FATAL is already recognised other than just at the beginning. For example, in 5.30.3 ...

```
$ perl -E 'use warnings qw(once FATAL redefine); $once=""; $SIG{__DIE__} = sub { print "Fatal\n" }; sub foo {} *foo = sub {0}'
Name "main::once" used only once: possible typo at -e line 1.
Fatal
Subroutine main::foo redefined at -e line 1.
$ echo $?
255

$ perl -E 'use warnings qw(once redefine); $once=""; $SIG{__DIE__} = sub { print "Fatal\n" }; sub foo {} *foo = sub {0}'
Name "main::once" used only once: possible typo at -e line 1.
Subroutine main::foo redefined at -e line 1.
$ echo $?
0
```

This behaviour is unchanged in this branch.

Grinnz commented on lib/warnings.pm:
```diff
@@ -337,14 +337,22 @@ sub import
+            my $mask = ${^WARNING_BITS} // ($^W ? $Bits{all} : $DEFAULT) ;
+            ${^WARNING_BITS} = _bits($mask, @fatal, $warning);
+        } else {
+            unimport(undef, substr($warning, 1))
```
`unimport` should get called as a method on the same invocant as `import`.

Grinnz commented:
One more comment: the description of this feature *must* specify what version of Perl it is first available in. Unfortunately there is some uncertainty as to what the next major version of Perl will be so we may need a placeholder version.

DrHyde commented:
> One more comment: the description of this feature _must_ specify what version of Perl it is first available in. Unfortunately there is some uncertainty as to what the next major version of Perl will be so we may need a placeholder version.

It won't be in the documentation for 5.32 and earlier, it will be documented and available in whatever the next version is, surely. But in any case, other changes in warnings are documented with a version number only when existing categories were updated, and when 'FATAL' on its own was changed to imply 'all'.

Grinnz commented:
> > One more comment: the description of this feature _must_ specify what version of Perl it is first available in. Unfortunately there is some uncertainty as to what the next major version of Perl will be so we may need a placeholder version.
> 
> It won't be in the documentation for 5.32 and earlier, it will be documented and available in whatever the next version is, surely. But in any case, other changes in warnings are documented with a version number only when existing categories were updated, and when 'FATAL' on its own was changed to imply 'all'.

Irrespective of precedent, this must be established in the documentation to make clear that one cannot rely on this feature on older versions, since it would cause compilation failure in such cases.

Grinnz commented:
I'll add, this is always useful information to include directly in the documentation for end users in such a case, and has been neglected far too often in the past.

DrHyde commented:
Unless someone comes up with a better idea for the placeholder, '5.34' seems suitable.

Grinnz commented on lib/warnings.pm:
```diff
@@ -658,6 +669,41 @@ be reported for the C<$x> variable.
 
+=head2 "Negative warnings"
+
+As a convenience, you can as of perl 5.34 pass arguments to the
```
Perl should be capitalized.

DrHyde commented:
Anything else needed? I'm pretty darned sure that I'm Not Guilty for the mingw64 test failures!

FGasper commented on regen/warnings.pl:
```diff
@@ -962,6 +973,41 @@ =head2 Default Warnings and Optional Warnings
 
+=head2 "Negative warnings"
+
+As a convenience, you can as of Perl 5.34 pass arguments to the
```
Grammatical nit: I think “as of Perl 5.34” should have commas around it.

(Same applies in warnings.pm.)

tom-binary replied:
that would, I suggest, lead, unintentionally, to an unfortunate case of [shatner comma](https://latimesblogs.latimes.com/jacketcopy/2011/06/goodbye-oxford-comma-hello-shatner-comma.html) - previous text seemed fine, but if you really want to break it up then why not `(as of Perl 5.34)`.

FGasper replied:
> 3. Use a pair of commas in the middle of a sentence to set off clauses, phrases, and words that are not essential to the meaning of the sentence.

https://owl.purdue.edu/owl/general_writing/punctuation/commas/extended_rules_for_commas.html

Parens would be a nice alternative, too.

DrHyde replied:
I do realise that in the time I've taken to type this I could have made the changes and committed them. I haven't, because the aim of review is to *improve* a PR, not to impose some individual's preferences. I'm sure that mechanically following those rules is good for non-native speakers wishing to produce clear if somewhat stilted prose, but that's all that mechanically following rules would be good for.

As a native English speaker for the past 40+ years I say with utter conviction that extra commas are unnecessary there. A splattering of extra commas wouldn't change the meaning in any way and, just like @tom-binary's example, can only, in this case, harm comprehensibility, just a little bit, by breaking up the flow.

FGasper replied:
@DrHyde The reason why I posted the link in my last response was to illustrate that this is not an individual preference. It’s just standard English.

Grinnz replied:
As a native English speaker, I agree that commas unnecessarily break the flow but parentheses improve it greatly.

FGasper replied:
As yet another native English speaker :), I agree that parens would be an improvement, probably better than the (otherwise requisite) commas.

FGasper commented on regen/warnings.pl:
```diff
@@ -962,6 +973,41 @@ =head2 Default Warnings and Optional Warnings
+
+As a convenience, you can as of Perl 5.34 pass arguments to the
+C<import()> method both positively and negatively. Negative warnings
+are those with a C<-> sign prepended to their names, positive warnings
```
Grammatical nit: Use a semicolon here rather than a comma.

(Same applies in warnings.pm.)

DrHyde replied:
Why? How would an extra pixel clarify the meaning?

FGasper replied:
I wouldn’t say it clarifies meaning; it’s just convention in written English that if you have 2 independent clauses—as is the case here—then use either a comma plus conjunction, or a semicolon.

So, either:
> … prepended to their names, and positive warnings …

… or:
> … prepended to their names; positive warnings …

FGasper commented on regen/warnings.pl:
```diff
@@ -962,6 +973,41 @@ =head2 Default Warnings and Optional Warnings
+        ...
+    }
+
+The argument list is processed in the order you specify. So for example if you
```
Grammatical nit: “for example” might be better with commas around it.

(Same applies in warnings.pm.)

DrHyde commented:
I care more about being able to use this code than about being right :-)

Grinnz reviewed (approved):
LGTM, needs squashing

DrHyde commented:
You mean just a big `git rebase -i` down to one commit?

khwilliamson commented:
I just looked at the commits, and in my opinion a single commit would be fine.  But lets give it a day or so to get other view points, if any.

DrHyde commented:
All squashed down to one commit
